// pixijs/webworker/filters.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('pixijs/core');
var filterAlpha = require('pixijs/filter/alpha');
var filterBlur = require('pixijs/filter/blur');
var filterColorMatrix = require('pixijs/filter/color-matrix');
var filterDisplacement = require('pixijs/filter/displacement');
var filterFxaa = require('pixijs/filter/fxaa');
var filterNoise = require('pixijs/filter/noise');

const filters = {
  AlphaFilter: filterAlpha.AlphaFilter,
  BlurFilter: filterBlur.BlurFilter,
  BlurFilterPass: filterBlur.BlurFilterPass,
  ColorMatrixFilter: filterColorMatrix.ColorMatrixFilter,
  DisplacementFilter: filterDisplacement.DisplacementFilter,
  FXAAFilter: filterFxaa.FXAAFilter,
  NoiseFilter: filterNoise.NoiseFilter
};
Object.entries(filters).forEach(([key, FilterClass]) => {
  Object.defineProperty(filters, key, {
    get() {
      core.utils.deprecation("7.1.0", `filters.${key} has moved to ${key}`);
      return FilterClass;
    }
  });
});

exports.filters = filters;
