// pixijs/node/index.mjs
import { ResizePlugin } from 'pixijs/app';
export * from 'pixijs/app';
import { loadTextures, loadWebFont } from 'pixijs/assets';
export * from 'pixijs/assets';
import { extensions, INSTALLED } from 'pixijs/core';
export * from 'pixijs/core';
export { settings } from 'pixijs/core';
import 'pixijs/display/cache-as-bitmap';
import 'pixijs/display/get-child-by-name';
import 'pixijs/display/get-global-position';
import './adapter/index.mjs';
export { filters } from './filters.mjs';
export * from 'pixijs/display';
export * from 'pixijs/extract';
export * from 'pixijs/filter/alpha';
export * from 'pixijs/filter/blur';
export * from 'pixijs/filter/color-matrix';
export * from 'pixijs/filter/displacement';
export * from 'pixijs/filter/fxaa';
export * from 'pixijs/filter/noise';
export * from 'pixijs/graphics';
export * from 'pixijs/mesh';
export * from 'pixijs/mesh/extras';
export * from 'pixijs/particle-container';
export * from 'pixijs/prepare';
export * from 'pixijs/sprite';
export * from 'pixijs/sprite-animated';
export * from 'pixijs/sprite-tiling';
export * from 'pixijs/spritesheet';
export * from 'pixijs/text';
export * from 'pixijs/text-bitmap';
import { NodeCanvasResource } from './adapter/NodeCanvasResource.mjs';
export { NodeCanvasResource } from './adapter/NodeCanvasResource.mjs';
export { NodeAdapter } from './adapter/adapter.mjs';
export { loadNodeBase64 } from './adapter/loadNodeBase64.mjs';
export { loadNodeFont } from './adapter/loadNodeFont.mjs';
export { loadNodeTexture } from './adapter/loadNodeTexture.mjs';
export { NodeCanvasElement } from './adapter/NodeCanvasElement.mjs';

extensions.remove(loadTextures, loadWebFont, ResizePlugin);
INSTALLED.length = 0;
INSTALLED.push(NodeCanvasResource);
